get-artifacts/README.md

```diff
@@ -0,0 +1,60 @@
+# Get Artifacts #
+
+A downloader for artifacts from multiple prior workflow_runs (by workflow_file_name and branch), or a single run by workflow_run_id.
+
+Artifacts are always uploaded as zips, correspondingly this action allows you to unzip and then delete the original downloads with the "decompress" flags.
+
+If an ```${{ inputs.target_dir }}/<workflow_run_id>/<artifact_name>/``` folder already exist no attempt will be made to download the artifacts for that run.  The assumption it github's cache action can be use to preserve prior downloads between workflow executions, preventing unnecessary downloads.  More info on cacheing [here](https://docs.github.com/en/actions/guides/caching-dependencies-to-speed-up-workflows "caching-dependencies-to-speed-up-workflows").
```

Comment:
`already exists,`

`The assumption is`

Reply:
I believe for the caching to work correctly (both for us and in general), concurrency must be limited to 1. Could the documentation be expanded to talk about it, and the example (and potentially CI) include that? https://docs.github.com/en/actions/reference/workflow-syntax-for-github-actions#concurrency

Reply:
I think the nature of github's caching is outside the scope of this.    there are ways to have more/less exact keys, and this script will fill in any holes it finds, so even if we ran it in two jobs at the same time both would end up with the full set of files.    Regardless this script is idempotent, so even if two jobs raced to build the prior cache, this would end up with a consistent set of files.   Effectively an eventually consistent algo for building the cache.

Reply:
I'm more worried about situations where you have two jobs A and B. A starts before B, downloads fewer files, but finishes after B (because of say network flakiness). I agree that ""eventually"" the full set of files will be filled out but I think situations like this are invariably going to catch people by surprise.

And for consumers, 99% of the time the correct thing to do would be to limit concurrency anyway -- for example, the flaky test detection job absolutely must be concurrency-limited because you don't want two concurrently-running jobs with slightly different views of the world trampling over each other.

I think the concurrency bit deserves a mention in the text at least.

Reply:
Oh one other question: how would cache cleanup work? I think anything older than HISTORY runs should be deleted?

Reply:
Being eventually consistent is a valid choice for use of artifacts, report A might have workflow alpha in it, and report B may have workflow beta, but may have workflow alpha.   However the next report C, will have both alpha and beta workflows since it will look for and download how every many historical workflows you ask for.

I don't think that is likely for our use case that this will ever occur,  since we will only run this when the auto branch completes and bors single threads that.

Cache cleanup would likely be a lexigraphical sort of the contents of the target dir and delete the oldest over some number.   I already have a similar shell in existence for older docker files.

But again, I contend that this is all outside the scope of the action.